### PR TITLE
Mirror entire instance

### PR DESCRIPTION
**Pull request by vadmeste**
Ready for first review

### CODE REVIEW

**Comment by vadmeste** [comment]
Ready for review

File: cmd/common-methods.go [hunk 1]
```diff
@@ -180,3 +181,16 @@ func newClient(aliasedURL string) (Client, *probe.Error) {
 	}
 	return newClientFromAlias(alias, urlStrFull)
 }
+
+// newS3Client gives S3 client if the url represents an S3 server
+func newS3Client(url string) (*s3Client, *probe.Error) {
```
Comment (harshavardhana) [inline cmd/common-methods.go]: why a function like this?

File: cmd/mirror-main.go [hunk 2]
```diff
@@ -52,6 +52,15 @@ var (
 			Name:  "remove",
 			Usage: "Remove extraneous file(s) on target.",
 		},
+		cli.StringFlag{
+			Name:  "region",
+			Usage: "Specify which region to select when creating new buckets.",
+			Value: "us-east-1",
+		},
+		cli.BoolFlag{
+			Name:  "preserve-policies, a",
```
Comment (harshavardhana) [inline cmd/mirror-main.go]: we don't need the long option just keep it "a"

File: cmd/mirror-main.go [hunk 3]
```diff
@@ -522,16 +516,148 @@ func newMirrorJob(ctx *cli.Context) *mirrorJob {
...
+		if d.Diff == differInFirst {
+			// Bucket only exists in the source, create the same bucket in the destination
+			err := newDstClt.MakeBucket(ctx.String("region"))
+			fatalIf(err, "Cannot created bucket in `"+newTgtURL+"`")
+		}
+
+		// Copy policy rules from source to dest if flag is activated
+		if ctx.Bool("preserve-policies") {
+			rules, err := newSrcClt.GetAccessRules()
+			fatalIf(err, "Cannot get policy rules")
```
Comment (harshavardhana) [inline cmd/mirror-main.go]: We don't need to fail we can proceed forward to the next one.

File: cmd/mirror-main.go [hunk 4]
```diff
@@ -522,16 +516,148 @@ func newMirrorJob(ctx *cli.Context) *mirrorJob {
...
+		newSrcURL := srcURL + d.firstContent.URL.Path
+		newTgtURL := dstURL + d.firstContent.URL.Path
+
+		newSrcClt, _ := newS3Client(newSrcURL)
+		newDstClt, _ := newS3Client(newTgtURL)
+
+		if d.Diff == differInFirst {
+			// Bucket only exists in the source, create the same bucket in the destination
+			err := newDstClt.MakeBucket(ctx.String("region"))
+			fatalIf(err, "Cannot created bucket in `"+newTgtURL+"`")
```
Comment (harshavardhana) [inline cmd/mirror-main.go]: Don't fail, proceed to next one.

File: cmd/mirror-main.go [hunk 5]
```diff
@@ -522,16 +516,148 @@ func newMirrorJob(ctx *cli.Context) *mirrorJob {
...
+		}
+
+		// Copy policy rules from source to dest if flag is activated
+		if ctx.Bool("preserve-policies") {
+			rules, err := newSrcClt.GetAccessRules()
+			fatalIf(err, "Cannot get policy rules")
+			// Set found rules to target bucket if permitted
+			for _, r := range rules {
+				originalRule, err := newDstClt.GetAccess()
+				fatalIf(err, "Cannot get bucket policy rules from `"+newTgtURL+"`")
```
Comment (harshavardhana) [inline cmd/mirror-main.go]: Don't fail proceed to next one.

File: cmd/mirror-main.go [hunk 6]
```diff
@@ -522,16 +516,148 @@ func newMirrorJob(ctx *cli.Context) *mirrorJob {
...
+
+	wg.Wait()
+	close(work)
+
+	return nil
+}
+
+// mirrorSingleBucket
+func mirrorSingleBucket(srcURL, tgtURL string, ctx *cli.Context) error {
+	mj := newMirrorJob(srcURL, tgtURL, 1,
```
Comment (harshavardhana) [inline cmd/mirror-main.go]: Even in single bucket if "-a" is specified should be honored.

File: cmd/mirror-main.go [hunk 7]
```diff
@@ -522,16 +516,148 @@ func newMirrorJob(ctx *cli.Context) *mirrorJob {
...
+
+	dstClt, err := newClient(tgtURL)
+	fatalIf(err, "Unable to initialize `"+tgtURL+"`")
+
+	mirrorAll := false
+
+	// Check if passed arguments could trigger a complete mirror instance
+	if srcClt.GetURL().Type == objectStorage && dstClt.GetURL().Type == objectStorage {
+		// Check if buckets are not passed
+		srcBucket, _ := srcClt.(*s3Client).url2BucketAndObject()
```
Comment (harshavardhana) [inline cmd/mirror-main.go]: We should never have code like this.. unless really really needed. what problem are you trying to solve here?

Comment (vadmeste) [inline cmd/mirror-main.go]: > We should never have code like this.. unless really really needed. what problem are you trying to solve here?

Here, I needed to check if the user passed any buckets in arguments to know if we are going to mirror all buckets or it is just a simple mirror

File: cmd/common-methods.go (hunk 1 above)
Comment (vadmeste) [inline cmd/common-methods.go]: > why a function like this?

Simplification purpose, this returns an error when this is not an s3 url.

**Comment by vadmeste** [comment]
@harshavardhana, I rewrite some codes after you gave some comments, but I didn't address all of them until we discuss them

File: cmd/common-methods.go (hunk 1 above)
Comment (harshavardhana) [inline cmd/common-methods.go]: Actually this is spilling the abstraction ..

Comment (vadmeste) [inline cmd/common-methods.go]: > Actually this is spilling the abstraction ..

You mean mirroring instance can be somehow abstracted too ?

**Pull request merged**